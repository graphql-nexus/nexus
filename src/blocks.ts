export { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
export { InterfaceDefinitionBlock } from './definitions/interfaceType'
export { ObjectDefinitionBlock } from './definitions/objectType'
export { UnionDefinitionBlock } from './definitions/unionType'
