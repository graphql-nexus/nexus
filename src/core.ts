// The "core" is used as a namespace to re-export everything,
// For anyone who wants to use the internals
export * from './builder'
export * from './makeSchema'
export * from './definitions/args'
export * from './definitions/decorateType'
export * from './definitions/definitionBlocks'
export * from './definitions/enumType'
export * from './definitions/extendInputType'
export * from './definitions/extendType'
export * from './definitions/inputObjectType'
export * from './definitions/interfaceType'
export * from './definitions/list'
export * from './definitions/mutationField'
export * from './definitions/mutationType'
export * from './definitions/nonNull'
export * from './definitions/nullable'
export * from './definitions/objectType'
export * from './definitions/queryField'
export * from './definitions/queryType'
export * from './definitions/scalarType'
export * from './definitions/subscriptionField'
export * from './definitions/subscriptionType'
export * from './definitions/nexusMeta'
export * from './definitions/unionType'
export * from './definitions/wrapping'
export * from './definitions/_types'
export * from './dynamicMethod'
export * from './plugin'
export * from './plugins'
export * from './rebuildType'
export * from './sdlConverter'
export * from './typegenAutoConfig'
export * from './typegenFormatPrettier'
export * from './typegenMetadata'
export * from './typegenPrinter'
export * from './typegenTypeHelpers'
export * from './typegenUtils'
export * from './utils'
