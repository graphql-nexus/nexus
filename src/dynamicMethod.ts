import { PluginBuilderLens } from './builder'
import { InputDefinitionBlock, OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'

export type OutputFactoryConfig<T> = {
  stage: 'walk' | 'build'
  args: any[]
  builder: PluginBuilderLens
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export type InputFactoryConfig<T> = {
  args: any[]
  builder: PluginBuilderLens
  typeDef: InputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface BaseExtensionConfig<T extends string> {
  /**
   * The name of the "extension", the field made
   * available on the builders
   */
  name: T
  /**
   * The full type definition for the options, including generic
   * signature for the type
   */
  typeDefinition?: string
}

export interface DynamicOutputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: OutputFactoryConfig<T>): any
}

export interface DynamicInputMethodConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the field is called
   */
  factory(config: InputFactoryConfig<T>): any
}

export class DynamicInputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicInputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicInputMethodDef, NexusTypes.DynamicInput)

export class DynamicOutputMethodDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputMethodConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputMethodDef, NexusTypes.DynamicOutputMethod)

/**
 * Defines a new property on the object definition block
 * for an output type, taking arbitrary input to define
 * additional types. See the connectionPlugin:
 *
 * t.connectionField('posts', {
 *   nullable: true,
 *   totalCount(root, args, ctx, info) {
 *     return ctx.user.getTotalPostCount(root.id, args)
 *   },
 *   nodes(root, args, ctx, info) {
 *     return ctx.user.getPosts(root.id, args)
 *   }
 * })
 */
export function dynamicOutputMethod<T extends string>(config: DynamicOutputMethodConfig<T>) {
  return new DynamicOutputMethodDef(config.name, config)
}

/**
 * Same as the outputFieldExtension, but for fields that
 * should be added on as input types.
 */
export function dynamicInputMethod<T extends string>(config: DynamicInputMethodConfig<T>) {
  return new DynamicInputMethodDef(config.name, config)
}
