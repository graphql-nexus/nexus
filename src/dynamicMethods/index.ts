export * from './collection'
export * from './relayConnection'
