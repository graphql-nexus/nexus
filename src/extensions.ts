import { GraphQLNamedType } from 'graphql'
import { DynamicFieldDefs, SchemaConfig } from './builder'
import { NexusOutputFieldConfig } from './definitions/definitionBlocks'
import { NexusInputObjectTypeConfig } from './definitions/inputObjectType'
import { NexusInterfaceTypeConfig } from './definitions/interfaceType'
import { NexusObjectTypeConfig } from './definitions/objectType'
import { RootTypings } from './definitions/_types'

export type NexusGraphQLNamedType = GraphQLNamedType & {
  extensions?: {
    nexus?: {
      config: any
    }
  }
}

export type NexusTypeExtensions = NexusObjectTypeExtension | NexusInterfaceTypeExtension

/**
 * Container object living on `fieldDefinition.extensions.nexus`
 */
export class NexusFieldExtension<TypeName extends string = any, FieldName extends string = any> {
  readonly config: Omit<NexusOutputFieldConfig<TypeName, FieldName>, 'resolve'>
  /**
   * Whether the user has provided a custom "resolve" function,
   * or whether we're using GraphQL's defaultResolver
   */
  readonly hasDefinedResolver: boolean
  constructor(config: NexusOutputFieldConfig<TypeName, FieldName>) {
    const { resolve, ...rest } = config
    this.config = rest
    this.hasDefinedResolver = Boolean(resolve)
  }
}

/**
 * Container object living on `inputObjectType.extensions.nexus`
 */
export class NexusInputObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInputObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInputObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `objectType.extensions.nexus`
 */
export class NexusObjectTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusObjectTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusObjectTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

/**
 * Container object living on `interfaceType.extensions.nexus`
 */
export class NexusInterfaceTypeExtension<TypeName extends string = any> {
  readonly config: Omit<NexusInterfaceTypeConfig<TypeName>, 'definition'>
  constructor(config: NexusInterfaceTypeConfig<TypeName>) {
    const { definition, ...rest } = config
    this.config = rest
  }
}

export interface NexusSchemaExtensionConfig extends Omit<SchemaConfig, 'types'> {
  dynamicFields: DynamicFieldDefs
  rootTypings: RootTypings
}

/**
 * Container object living on `schema.extensions.nexus`. Keeps track
 * of metadata from the builder so we can use it when we
 */
export class NexusSchemaExtension {
  constructor(readonly config: NexusSchemaExtensionConfig) {}
}
