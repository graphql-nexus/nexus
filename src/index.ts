import * as blocks from './blocks'
import * as core from './core'

// All of the Public API definitions
export { makeSchema } from './makeSchema'
export { arg, booleanArg, floatArg, idArg, intArg, stringArg } from './definitions/args'
export { decorateType } from './definitions/decorateType'
export { enumType } from './definitions/enumType'
export { extendInputType } from './definitions/extendInputType'
export { extendType } from './definitions/extendType'
export { inputObjectType } from './definitions/inputObjectType'
export { interfaceType } from './definitions/interfaceType'
export { list } from './definitions/list'
export { mutationField } from './definitions/mutationField'
export { mutationType } from './definitions/mutationType'
export { nonNull } from './definitions/nonNull'
export { nullable } from './definitions/nullable'
export { objectType } from './definitions/objectType'
export { queryField } from './definitions/queryField'
export { queryType } from './definitions/queryType'
export { asNexusMethod, scalarType } from './definitions/scalarType'
export { subscriptionField } from './definitions/subscriptionField'
export { subscriptionType } from './definitions/subscriptionType'
export { unionType } from './definitions/unionType'
export { dynamicInputMethod, dynamicOutputMethod } from './dynamicMethod'
export { dynamicOutputProperty } from './dynamicProperty'
export { createPlugin, plugin, PluginBuilderLens, PluginConfig } from './plugin'
export * from './plugins'
export { convertSDL } from './sdlConverter'
export { AllInputTypes, AllOutputTypes, FieldResolver, FieldType } from './typegenTypeHelpers'
export { groupTypes } from './utils'
export { blocks, core }
