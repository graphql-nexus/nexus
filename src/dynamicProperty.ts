import { SchemaBuilder } from './builder'
import { OutputDefinitionBlock } from './definitions/definitionBlocks'
import { NexusTypes, withNexusSymbol } from './definitions/_types'
import { BaseExtensionConfig } from './dynamicMethod'

export type OutputPropertyFactoryConfig<T> = {
  stage: 'walk' | 'build'
  builder: SchemaBuilder
  typeDef: OutputDefinitionBlock<any>
  /**
   * The name of the type this field is being declared on
   */
  typeName: string
}

export interface DynamicOutputPropertyConfig<T extends string> extends BaseExtensionConfig<T> {
  /**
   * Invoked when the property is accessed (as a getter)
   */
  factory(config: OutputPropertyFactoryConfig<T>): any
}

export class DynamicOutputPropertyDef<Name extends string> {
  constructor(readonly name: Name, protected config: DynamicOutputPropertyConfig<Name>) {}
  get value() {
    return this.config
  }
}
withNexusSymbol(DynamicOutputPropertyDef, NexusTypes.DynamicOutputProperty)

/**
 * Defines a new property on the object definition block
 * for an output type, making it possible to build custom DSL's
 * on top of Nexus, e.g. in nexus-prisma
 *
 * t.model.posts()
 */
export function dynamicOutputProperty<T extends string>(config: DynamicOutputPropertyConfig<T>) {
  return new DynamicOutputPropertyDef(config.name, config)
}
