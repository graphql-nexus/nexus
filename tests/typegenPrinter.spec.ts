import { buildSchema, GraphQLField, GraphQLInterfaceType, GraphQLObjectType } from 'graphql'
import * as path from 'path'
import { core } from '../src'
import { EXAMPLE_SDL } from './_sdl'
const { makeSchema, TypegenPrinter, TypegenMetadata } = core

describe('typegenPrinter', () => {
  let typegen: core.TypegenPrinter
  let metadata: core.TypegenMetadata
  beforeEach(async () => {
    const schema = makeSchema({
      outputs: {
        typegen: path.join(__dirname, 'typegen/types.gen.ts'),
        schema: path.join(__dirname, 'typegen/schema.gen.graphql'),
      },
      shouldGenerateArtifacts: true,
      types: [buildSchema(EXAMPLE_SDL)],
      prettierConfig: path.join(__dirname, '../package.json'),
    }) as core.NexusGraphQLSchema
    metadata = new TypegenMetadata({
      outputs: {
        typegen: path.join(__dirname, 'test-gen.ts'),
        schema: path.join(__dirname, 'test-gen.graphql'),
      },
      typegenAutoConfig: {
        backingTypeMap: {
          UUID: 'string',
        },
        sources: [
          {
            alias: 't',
            source: path.join(__dirname, '__helpers/index.ts'),
          },
        ],
        contextType: 't.TestContext',
      },
    })
    // give time for artifact generation to complete
    await new Promise((res) => setTimeout(res, 2000))
    const typegenInfo = await metadata.getTypegenInfo(schema)
    typegen = new TypegenPrinter(metadata.sortSchema(schema), {
      ...typegenInfo,
      typegenFile: '',
    })
    jest
      .spyOn(typegen, 'hasResolver')
      .mockImplementation((field: GraphQLField<any, any>, type: GraphQLObjectType | GraphQLInterfaceType) => {
        if (type.name === 'Query' || type.name === 'Mutation') {
          return true
        }
        return false
      })
  })

  it('builds the enum object type defs', () => {
    expect(typegen.printEnumTypeMap()).toMatchSnapshot()
  })

  it('builds the input object type defs', () => {
    expect(typegen.printInputTypeMap()).toMatchSnapshot()
  })

  it('should build an argument type map', () => {
    expect(typegen.printArgTypeMap()).toMatchSnapshot()
  })

  it('should print a root type map', () => {
    expect(typegen.printRootTypeMap()).toMatchSnapshot()
  })

  it('should not print roots for fields with resolvers', () => {
    // If we don't have a dedicated "root type", then we need to infer
    // what the return type should be based on the shape of the object
    // If the field has a resolver, we assume it's derived, otherwise
    // you'll need to supply a backing root type with more information.
    jest
      .spyOn(typegen, 'hasResolver')
      .mockImplementation((field: GraphQLField<any, any>, type: GraphQLObjectType | GraphQLInterfaceType) => {
        if (type.name === 'Query' || type.name === 'Mutation') {
          return true
        }
        if (type.name === 'User' && field.name === 'posts') {
          return true
        }
        return false
      })
    expect(typegen.printRootTypeMap()).toMatchSnapshot()
  })

  it('should print a return type map', () => {
    expect(typegen.printFieldTypesMap()).toMatchSnapshot()
  })

  it('should print the full output', () => {
    expect(typegen.print()).toMatchSnapshot()
  })
})
