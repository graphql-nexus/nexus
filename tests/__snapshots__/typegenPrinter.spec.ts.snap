// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typegenPrinter builds the enum object type defs 1`] = `
"export interface NexusGenEnums {
  OrderEnum: \\"ASC\\" | \\"DESC\\"
  SomeEnum: \\"A\\" | \\"B\\"
}"
`;

exports[`typegenPrinter builds the input object type defs 1`] = `
"export interface NexusGenInputs {
  CreatePostInput: { // input type
    author: string; // ID!
    geo: Array<Array<number | null>>; // [[Float]!]!
    name: string; // String!
  }
  PostFilters: { // input type
    order: NexusGenEnums['OrderEnum']; // OrderEnum!
    search: string | null; // String
  }
}"
`;

exports[`typegenPrinter should build an argument type map 1`] = `
"export interface NexusGenArgTypes {
  Mutation: {
    createPost: { // args
      input: NexusGenInputs['CreatePostInput']; // CreatePostInput!
    }
    registerClick: { // args
      uuid?: NexusGenScalars['UUID'] | null; // UUID
    }
    someList: { // args
      items: Array<string | null>; // [String]!
    }
  }
  Query: {
    posts: { // args
      filters: NexusGenInputs['PostFilters']; // PostFilters!
    }
  }
  User: {
    name: { // args
      prefix?: string | null; // String
    }
    posts: { // args
      filters?: NexusGenInputs['PostFilters'] | null; // PostFilters
    }
  }
}"
`;

exports[`typegenPrinter should not print roots for fields with resolvers 1`] = `
"export interface NexusGenObjects {
  Mutation: {};
  Post: { // root type
    author: NexusGenRootTypes['User']; // User!
    geo: number[][]; // [[Float!]!]!
    id: string; // ID!
    messyGeo?: Array<number[] | null> | null; // [[Float!]]
    uuid: NexusGenScalars['UUID']; // UUID!
  }
  Query: {};
  User: { // root type
    email: string; // String!
    id: string; // ID!
    name: string; // String!
    outEnum?: NexusGenEnums['SomeEnum'] | null; // SomeEnum
    phone?: string | null; // String
  }
}"
`;

exports[`typegenPrinter should not print roots for fields with resolvers 2`] = `"export type NexusGenRootTypes = NexusGenInterfaces & NexusGenObjects & NexusGenUnions"`;

exports[`typegenPrinter should print a interface type map 1`] = `
"export interface NexusGenInterfaces {
  Node: core.Discriminate<'Post', 'required'> | core.Discriminate<'User', 'required'>;
}"
`;

exports[`typegenPrinter should print a object type map 1`] = `
"export interface NexusGenObjects {
  Mutation: {};
  Post: { // root type
    author: NexusGenRootTypes['User']; // User!
    geo: number[][]; // [[Float!]!]!
    id: string; // ID!
    messyGeo?: Array<number[] | null> | null; // [[Float!]]
    uuid: NexusGenScalars['UUID']; // UUID!
  }
  Query: {};
  User: { // root type
    email: string; // String!
    id: string; // ID!
    name: string; // String!
    outEnum?: NexusGenEnums['SomeEnum'] | null; // SomeEnum
    phone?: string | null; // String
    posts: NexusGenRootTypes['Post'][]; // [Post!]!
  }
}"
`;

exports[`typegenPrinter should print a return type map 1`] = `
"export interface NexusGenFieldTypes {
  Mutation: { // field return type
    createPost: NexusGenRootTypes['Post']; // Post!
    registerClick: NexusGenRootTypes['Query']; // Query!
    someList: Array<string | null>; // [String]!
  }
  Post: { // field return type
    author: NexusGenRootTypes['User']; // User!
    geo: number[][]; // [[Float!]!]!
    id: string; // ID!
    messyGeo: Array<number[] | null> | null; // [[Float!]]
    uuid: NexusGenScalars['UUID']; // UUID!
  }
  Query: { // field return type
    posts: NexusGenRootTypes['Post'][]; // [Post!]!
    unionField: NexusGenRootTypes['ExampleUnion']; // ExampleUnion!
    user: NexusGenRootTypes['User']; // User!
  }
  User: { // field return type
    email: string; // String!
    id: string; // ID!
    name: string; // String!
    outEnum: NexusGenEnums['SomeEnum'] | null; // SomeEnum
    phone: string | null; // String
    posts: NexusGenRootTypes['Post'][]; // [Post!]!
  }
  Node: { // field return type
    id: string; // ID!
  }
}"
`;

exports[`typegenPrinter should print a root type map 1`] = `"export type NexusGenRootTypes = NexusGenInterfaces & NexusGenObjects & NexusGenUnions"`;

exports[`typegenPrinter should print a union type map 1`] = `
"export interface NexusGenUnions {
  ExampleUnion: core.Discriminate<'Post', 'required'> | core.Discriminate<'User', 'required'>;
}"
`;

exports[`typegenPrinter should print the full output 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */

import * as t from \\"./__helpers/index\\"
import { core } from \\"@nexus/schema\\"




declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
  CreatePostInput: { // input type
    author: string; // ID!
    geo: Array<Array<number | null>>; // [[Float]!]!
    name: string; // String!
  }
  PostFilters: { // input type
    order: NexusGenEnums['OrderEnum']; // OrderEnum!
    search: string | null; // String
  }
}

export interface NexusGenEnums {
  OrderEnum: \\"ASC\\" | \\"DESC\\"
  SomeEnum: \\"A\\" | \\"B\\"
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  UUID: string
}

export interface NexusGenObjects {
  Mutation: {};
  Post: { // root type
    author: NexusGenRootTypes['User']; // User!
    geo: number[][]; // [[Float!]!]!
    id: string; // ID!
    messyGeo?: Array<number[] | null> | null; // [[Float!]]
    uuid: NexusGenScalars['UUID']; // UUID!
  }
  Query: {};
  User: { // root type
    email: string; // String!
    id: string; // ID!
    name: string; // String!
    outEnum?: NexusGenEnums['SomeEnum'] | null; // SomeEnum
    phone?: string | null; // String
    posts: NexusGenRootTypes['Post'][]; // [Post!]!
  }
}

export interface NexusGenInterfaces {
  Node: core.Discriminate<'Post', 'required'> | core.Discriminate<'User', 'required'>;
}

export interface NexusGenUnions {
  ExampleUnion: core.Discriminate<'Post', 'required'> | core.Discriminate<'User', 'required'>;
}

export type NexusGenRootTypes = NexusGenInterfaces & NexusGenObjects & NexusGenUnions

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars & NexusGenEnums

export interface NexusGenFieldTypes {
  Mutation: { // field return type
    createPost: NexusGenRootTypes['Post']; // Post!
    registerClick: NexusGenRootTypes['Query']; // Query!
    someList: Array<string | null>; // [String]!
  }
  Post: { // field return type
    author: NexusGenRootTypes['User']; // User!
    geo: number[][]; // [[Float!]!]!
    id: string; // ID!
    messyGeo: Array<number[] | null> | null; // [[Float!]]
    uuid: NexusGenScalars['UUID']; // UUID!
  }
  Query: { // field return type
    posts: NexusGenRootTypes['Post'][]; // [Post!]!
    unionField: NexusGenRootTypes['ExampleUnion']; // ExampleUnion!
    user: NexusGenRootTypes['User']; // User!
  }
  User: { // field return type
    email: string; // String!
    id: string; // ID!
    name: string; // String!
    outEnum: NexusGenEnums['SomeEnum'] | null; // SomeEnum
    phone: string | null; // String
    posts: NexusGenRootTypes['Post'][]; // [Post!]!
  }
  Node: { // field return type
    id: string; // ID!
  }
}

export interface NexusGenFieldTypeNames {
  Mutation: { // field return type name
    createPost: 'Post'
    registerClick: 'Query'
    someList: 'String'
  }
  Post: { // field return type name
    author: 'User'
    geo: 'Float'
    id: 'ID'
    messyGeo: 'Float'
    uuid: 'UUID'
  }
  Query: { // field return type name
    posts: 'Post'
    unionField: 'ExampleUnion'
    user: 'User'
  }
  User: { // field return type name
    email: 'String'
    id: 'ID'
    name: 'String'
    outEnum: 'SomeEnum'
    phone: 'String'
    posts: 'Post'
  }
  Node: { // field return type name
    id: 'ID'
  }
}

export interface NexusGenArgTypes {
  Mutation: {
    createPost: { // args
      input: NexusGenInputs['CreatePostInput']; // CreatePostInput!
    }
    registerClick: { // args
      uuid?: NexusGenScalars['UUID'] | null; // UUID
    }
    someList: { // args
      items: Array<string | null>; // [String]!
    }
  }
  Query: {
    posts: { // args
      filters: NexusGenInputs['PostFilters']; // PostFilters!
    }
  }
  User: {
    name: { // args
      prefix?: string | null; // String
    }
    posts: { // args
      filters?: NexusGenInputs['PostFilters'] | null; // PostFilters
    }
  }
}

export interface NexusGenAbstractTypeMembers {
  ExampleUnion: \\"Post\\" | \\"User\\"
  Node: \\"Post\\" | \\"User\\"
}

export interface NexusGenTypeInterfaces {
  Post: \\"Node\\"
  User: \\"Node\\"
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = keyof NexusGenInputs;

export type NexusGenEnumNames = keyof NexusGenEnums;

export type NexusGenInterfaceNames = keyof NexusGenInterfaces;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = keyof NexusGenUnions;

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    __typename: true
    isTypeOf: false
    resolveType: false
  }
}

export interface NexusGenTypes {
  context: t.TestContext;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginSchemaConfig {
  }
}"
`;
